src/rcl_lifecycle_bindings.cpp

```diff
@@ -0,0 +1,372 @@
+
+  RclHandle* state_machine_handle = RclHandle::Unwrap<RclHandle>(
+      Nan::To<v8::Object>(info[1]).ToLocalChecked());
+  rcl_lifecycle_state_machine_t* state_machine =
+      reinterpret_cast<rcl_lifecycle_state_machine_t*>(
+          state_machine_handle->ptr());
+
+  std::string lifecycle_srv_field_name(*Nan::Utf8String(info[2]));
+
+  std::string service_name;
+  rcl_service_t* service;
+  if (lifecycle_srv_field_name.compare("srv_get_state") == 0) {
+    service_name = "~/get_state";
+    service = &state_machine->com_interface.srv_get_state;
+  } else if (lifecycle_srv_field_name.compare("srv_get_available_states") ==
+             0) {
+    service_name = "~/get_available_states";
+    service = &state_machine->com_interface.srv_get_available_states;
+  } else if (lifecycle_srv_field_name.compare(
+                 "srv_get_available_transitions") == 0) {
+    service_name = "~/get_available_transitions";
+    service = &state_machine->com_interface.srv_get_available_transitions;
+  } else if (lifecycle_srv_field_name.compare("srv_change_state") == 0) {
+    service_name = "~/change_state";
+    service = &state_machine->com_interface.srv_change_state;
+  }
+
+  // build result object {name: <srv_name>, handle: <rcl handle of service_t>}
+  v8::Local<v8::Object> named_srv_obj = Nan::New<v8::Object>();
+  Nan::Set(named_srv_obj, Nan::New("name").ToLocalChecked(),
+           Nan::New(service_name).ToLocalChecked());
+
+  // Note: the following hack is to avoid a 'double free()' error that arises
+  // when overlapping RclHandles are created for a rcl_lifecycle_state_model and
+  // for each of the state-model's 4 services. The deleter for the state-model
+  // RclHandle calls rcl_lifecycle_state_machine_fini(state-model) which frees
+  // each of the state-model services. Then the RclHandle deleter for each
+  // state-model service is invoked resulting in a call to free() on the already
+  // freed service ptr.
+  // *** Boom - double free error ***.
+  //
+  // The work-around (hack) is to prevent the RclHandlers wrapping each of the
+  // state-machine services from calling free() on their service ptr. This is
+  // accomplished by setting each of the service RclHandlers' pointer field to
+  // nullptr causing their Reset() method to return immediately when invoked.
+  //
+  // The implementation of this approach replaces the nop deleter function on
+  // each service RclHandler with a new function that is passed a pointer to the
+  // service RclHandler. The deleter function then sets the service RclHandler
+  // parameter's pointer to nullptr. All joy - no double free()!
+  auto srv_handle =
+      RclHandle::NewInstance(service, nullptr, [] { return RCL_RET_OK; });
+  RclHandle* rclHandle = RclHandle::Unwrap<RclHandle>(srv_handle);
+  RclHandle::Unwrap<RclHandle>(srv_handle)->set_deleter([rclHandle] {
+      rclHandle->set_ptr(nullptr);
+      return RCL_RET_OK;
+    });
```

Comment:
I would appreciate a close critique of this unit of code where I worked around an exception where free() was being called twice on the same underlying rcl unit of memory. In a nutshell, rcl provides a rcl_lifecycle_state_model that contains a state-model and also instances of Services for the lifecycle_msgs/srv/* interfaces. The LifecycleNode implementation accesses these Services from the rcl_lifecycle_state_model, creates RclHandles for each and then installs them in the node's services list. When the node is gc'ed the deleter for the RclHandle wrapping the rcl_lifecycle_state_model properly cleans up it's memory as well as releasing the Services it manages. The RclHandles wrapping each of the state_model services are calling free() on the service pointer which has already been freed by it's parent state_model. This is causing a segfault.

To avoid this 2nd free() I'm hacking the service RclHandler deleter such that it sets it's pointer to a service to null. Doing this prevents RclHandle#Reset() from calling free().

Hope this makes sense.

Reply:
This is a new scenario of `RclHandle`, currently, the pointer was freed by default when calling `Reset()`. I suggest we could extend the `RclHandle` class to make the code consistent

For the declaration:
```c
  // |handle| will be freed if |will_delete| is true.
  static v8::Local<v8::Object> NewInstance(
      void* handle, RclHandle* parent = nullptr,
      std::function<int()> deleter = [] { return 0; },
      will_delete = true)
```

For the implementation:
```c
if (will_delete_)
  free(pointer_);
```

than we could call like
```c
 auto srv_handle =
      RclHandle::NewInstance(service, nullptr, std::function<int()>(), false);
```

Also, I notice that we have https://github.com/RobotWebTools/rclnodejs/blob/711072d689acc7221cbb8e289ca552db2c423c4b/src/rcl_handle.cpp#L81-L86, not sure if we could leverage that.

Reply:
I suggest to instead make the `deleter`'s job to free the memory, the default delete will call free but you can pass in a custom deleter to perform other actions.

Reply:
see issue #728